demux/packet.c

```diff
@@ -209,6 +211,30 @@ size_t demux_packet_estimate_total_size(struct demux_packet *dp)
     if (dp->avpacket) {
         mp_assert(!dp->is_cached);
         size += ROUND_ALLOC(dp->len);
+        if (dp->is_wrapped_avframe) {
+            mp_require(dp->buffer);
+
+            const AVFrame *frame = (AVFrame *)dp->buffer;
+            if (frame->hw_frames_ctx) {
+                const AVHWFramesContext *hwctx = (AVHWFramesContext *)frame->hw_frames_ctx->data;
+                size += ROUND_ALLOC(av_image_get_buffer_size(hwctx->sw_format, hwctx->width,
+                                                             hwctx->height, 16));
+            }
+            for (int i = 0; i < MP_ARRAY_SIZE(frame->buf) && frame->buf[i]; ++i)
+                size += ROUND_ALLOC(frame->buf[i]->size);
+
+            size += ROUND_ALLOC(frame->nb_extended_buf * sizeof(frame->extended_buf[0]));
+            for (int i = 0; i < frame->nb_extended_buf; ++i) {
+                size += ROUND_ALLOC(sizeof(*frame->extended_buf[i]));
+                size += ROUND_ALLOC(frame->extended_buf[i]->size);
+            }
+
+            size += ROUND_ALLOC(frame->nb_side_data * sizeof(frame->side_data[0]));
```

Comment:
How strict are we trying to be? There's also `frame->metadata` which is an `AVDictionary *` that contains more malloc'd data.

Reply:
In ideal world, very strict... but in practice, I think if we underestimate few bytes it will still work.

I think I would skip this, because iterating over dict and doing strlen of every key/value is not best here.

Also note that for AVFrame little difference in estimated size is not critical, because we won't cache huge amount of them, because singe frame is already quite big, so it quickly will fill the buffer limit.

Reply:
Personally I think we could skip the small stuff, but the existing code already does stuff like adding in `sizeof(pointer type)`

Reply:
If we're skipping small stuff then it's probably fine to ignore things like the `AVDictionary *metadata;` and the `AVChannelLayout layout;`

Reply:
It's kinda important for small packets. We can store millions of them, so it adds up quickly. But for big packets like frame here, the small error in size will not be significant.